app/proxy/proxy_ws.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2021 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package proxy
+
+// Inspired by https://github.com/koding/websocketproxy
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"github.com/gorilla/websocket"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	maxWsConnections = 1
+)
+
+var (
+	// From https://github.com/mendersoftware/deviceconnect/blob/master/api/http/device.go
+	upgrader = websocket.Upgrader{
+		Subprotocols: []string{"protomsg/msgpack"},
```

Comment:
This is the way the _server_ specifies which websocket subprotocols it supports. It cannot be fetched from the incoming headers because then you would be "supporting" whatever the client asks for. In Mender backend we only support `protomsg/msgpack` ATM.

Reply:
Discussed on Slack, will look into it, but outcome is still uncertain.

Reply:
Right, so what I have implemented is for the proxy:
* on client to backed direction, propagate `Sec-WebSocket-Protocol` header from the client (handshake request)
* on backend to client direction, support only `Sec-WebSocket-Protocol` header from the backend (handshake response)


```
CLIENT -----------------------> PROXY -----------------------> BACKEND
       protocol1, protocol 2          protocol1, protocol 2

CLIENT <----------------------- PROXY <----------------------- BACKEND
                      protocol1                      protocol1
```

I added one unit test and made also some local experiments. It should be fine

Reply:
Good, that's exactly what I meant, and what I hoped it would support!

Reply:
Yes! Thanks again for voicing this concern, I think we got it right now :tada: